src/lib/receipt_chain_database_lib/database.ml

```diff
@@ -0,0 +1,253 @@
+      Key_value_db.get t ~key:start_receipt
+      |> Result.of_option
+           ~error:
+             (Error.createf
+                !"Cannot retrieve proving receipt: %{sexp: \
+                  Receipt_chain_hash.t}"
+                start_receipt)
+      |> Or_error.bind ~f:(function
+           | Root _ ->
+               Or_error.errorf
+                 !"The transaction of root hash %{sexp: Receipt_chain_hash.t} \
+                   is not recorded"
+                 start_receipt
+           | Child {value; children; _} ->
+               if Receipt_chain_hash.equal start_receipt expected_receipt then
+                 Or_error.return [(start_receipt, value)]
+               else
+                 Sequence.of_list children
+                 |> Sequence.find_map ~f:(fun child ->
+                        match dfs acc child resulting_receipt with
+                        | Ok result -> Some ((start_receipt, value) :: result)
+                        | Error _ -> None )
+                 |> Result.of_option
+                      ~error:
+                        (Error.createf
+                           !"Could not find resulting receipt: %{sexp: \
+                             Receipt_chain_hash.t}"
+                           resulting_receipt) )
+    in
+    dfs [] proving_receipt resulting_receipt
+
+  let get_transaction t ~receipt =
+    Key_value_db.get t ~key:receipt
+    |> Option.bind ~f:(function
+         | Root _ -> None
+         | Child {value; _} -> Some value )
+
+  let add t ~previous (transaction: Transaction.t) =
+    let payload = Transaction.payload transaction in
+    let receipt_chain_hash = Receipt_chain_hash.cons payload previous in
+    let open Result.Let_syntax in
+    let result =
+      let%map children =
+        Option.value_map (Key_value_db.get t ~key:receipt_chain_hash)
+          ~default:(Ok []) ~f:(function
+          | Root {children} -> Ok children
+          | Child {parent= retrieved_parent; children; _} ->
+              if not (Receipt_chain_hash.equal previous retrieved_parent) then
+                Error (`Error_multiple_previous_receipts retrieved_parent)
+              else Ok children )
+      in
+      let node = Child {parent= previous; children; value= transaction} in
+      Key_value_db.set t ~key:receipt_chain_hash ~data:node ;
+      let updated_previous_node, status =
+        match Key_value_db.get t ~key:previous with
+        | None ->
+            (Root {children= [receipt_chain_hash]}, `Ok receipt_chain_hash)
+        | Some (Child previous_node) ->
+            ( Child
+                { previous_node with
+                  children= receipt_chain_hash :: previous_node.children }
+            , `Ok receipt_chain_hash )
+        | Some (Root previous_node) ->
+            ( Root {children= receipt_chain_hash :: previous_node.children}
+            , `Duplicate receipt_chain_hash )
+      in
+      Key_value_db.set t ~key:previous ~data:updated_previous_node ;
```

Comment:
what happens if the node dies between these two `set` calls? I think we're left in an inconsistent state. Since we never need to delete entries from the middle or end, maybe we don't need the previous hash? Then we don't have to worry about this case. Otherwise we should do `set_batch` or something (and make sure set_batch is a atomic)